api/odilo.py

```diff
@@ -0,0 +1,532 @@
+                error = data['errors'][0]
+                if error['id'] == 'CHECKOUT_NOT_FOUND':
+                    raise NotCheckedOut(error['description'])
+                else:
+                    raise CannotReturn(error['id'] + ': ' + error['description'])
+        else:
+            return response
+
+    @classmethod
+    def extract_date(cls, data, field_name):
+        if field_name not in data:
+            d = None
+        else:
+            # OdiloAPI dates are timestamps in milliseconds
+            d = datetime.datetime.fromtimestamp(float(data[field_name]) / 1000.0)
+        return d
+
+    def raise_exception_on_error(self, data, custom_error_to_exception={}):
+        if not data or 'errors' not in data or len(data['errors']) <= 0:
+            return
+
+        error = data['errors'][0]
+        error_code = error['id']
+        message = ('description' in error and error['description']) or ''
+        for d in custom_error_to_exception, self.error_to_exception:
+            if error in d:
+                raise d[error_code](message)
+
+    def get_loan(self, patron, pin, checkout_id):
+        url = self.CHECKOUT_GET.replace("{checkoutId}", checkout_id) + '?patronId=' + patron
+        response = self.patron_request(patron, pin, url)
+        if response.status_code == 200 and response.content:
+            data = response.json()
+            self.raise_exception_on_error(data)
+            return data
+        else:
+            return None
+
+    def get_hold(self, patron, pin, hold_id):
+        url = self.HOLD_GET.replace("{holdId}", hold_id)
+        data = self.patron_request(patron, pin, url).json()
+        self.raise_exception_on_error(data)
+        return data
+
+    def fulfill(self, patron, pin, licensepool, internal_format):
+        checkout_id = licensepool.identifier.identifier
+        content_link, content, content_type = self.get_fulfillment_link(patron, pin, checkout_id, internal_format)
+
+        if not content_link and not content:
+            self.log.info("Odilo record_id %s was not available as %s" % (checkout_id, internal_format))
+        else:
+            return FulfillmentInfo(
+                licensepool.collection,
+                DataSource.ODILO,
+                Identifier.ODILO_ID,
+                checkout_id,
+                content_link=content_link,
+                content=content,
+                content_type=content_type,
+                content_expires=None
+            )
+
+    def get_fulfillment_link(self, patron, pin, checkout_id, format_type):
+        """Get the link corresponding to an existing loan.
+        """
+        # Retrieve loan with its download_ulr. It is necessary to generate a download token in our API
+        loan = self.get_loan(patron, pin, checkout_id)
+        if not loan:
+            raise NoActiveLoan("Could not find active loan for %s" % checkout_id)
+
+        if format_type and loan['format'] and format_type == loan['format']:
```

Comment:
This whole thing would also benefit from being made into an independently testable method.